array.h

```diff
@@ -779,32 +892,38 @@ void resolve_unknown_strides(AllDims& all_dims, Dim0& dim0, Dims&... dims) {
 }
 
 template <class Dims, size_t... Is>
+NDARRAY_DECORATE_IF_CUDA
```

Comment:
I am not sure if this (and the above transitive functions up to `is_stride_ok`) are something that should be enabled on device. This happens when constructing an array/array_ref with a new shape, should that be something allowed in device code? (array for sure not, but array_ref?)

Reply:
Or maybe they should be overloaded on device to just assert all the strides are known instead of trying to dynamically fill them in? (`is_resolved` below.)

Reply:
Hm I guess it's up to philosophy. I was going for consistency: everything that can on device should. Then we won't have any code divergence.

Reply:
We should definitely allow constructing new array_refs on device. You might want to call another function. And in modern CUDA, you can launch a kernel from another kernel.

Reply:
I'm OK with leaving it as-is (enabling resolve() in __device__).

But FWIW, you would be able to construct a new array_ref, it would just require you to provide a shape without any unknown strides to be filled in by resolve.